ethereum-optimism/optimism
# op-node: generalize and improve alt-sync

**Description**

This PR iterates on top of the RPC-sync implementation to prepare for P2P-sync functionality as second form of alt-sync method, by encapsulating the RPC sync and updating its interface to make p2p sync fit in as well. Also fixes an off-by-one bug in the range to sync.

Changes:

- Add trust-RPC flag to `L2SyncEndpointConfig`. Previously it was hardcoded to true, yet also called blockhash verification. We can just allow the user to set it like they do with the L1 RPC.
- Change the sync-config check to be checked during the other config checks
- Clarify `GetUnsafeQueueGap` `start` is incl., `end` is excl., and fix off by one issue: we don't want to include the first block of the queue, but we do want to include the expected block number if queue is empty.
- Move RPC sync init code to its own function, separating it more from the l2 init
- Don't `Close` the RPC sync by reaching into attributes of the driver, but instead maintain the sync as separate lifetime thing in the node to close; there will be other sync methods.
- Put RPC sync requests behind an `AltSync` interface: the node itself implements it, and then routes it to the active & preferred sync methods that each also implement it, to allow more to be added.
  - The interface requests a full range of blocks, to avoid splitting the range into tinier tasks upfront, to allow the sync-method implementation to do batch / divide as desired.
  - The RPC `SyncClient` now consumes these range requests and does the fetching, instead of getting individual block numbers. Once new range information arrives, we now also drop previously scheduled information that may be stale.
- The driver now triggers sync sooner, when it's not changing the unsafe head as frequently as expected (thus not requesting syncing while it's deriving or has gossip coming in as expected). Also don't trigger sync when resetting the engine state (since we can't determine the range of L2 blocks accurately until the engine state is consistent again)

**Tests**

- Change RPC op-e2e test to disable batcher completely. When the sync runs around the same time as it syncs the data from L2, not all the blocks may be synced through the RPC. By disabling the batcher, we ensure it's really syncing through the RPC, and avoid potential flakiness.
- Change the same RPC test to use stringified block IDs, for more readable debug messages of what is and is not included in the received/published arrays
- Rename `L2SyncRPCConfig` to `PreparedL2SyncEndpoint` to match the other config utils of L1 and L2 endpoints that insert a prepared RPC client directly.


**Metadata**

Fix CLI-3643
